Update database.go

aleksandra0KR opened the pull request:
feat: registration with validity checks

aleksandra0KR committed:
Update database.go

feat: registration with validity checks

lilpipidron commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+	}
+	fmt.Fprint(w, "congrats, your account has been successfully created")
+
+}
```
Разбей на отдельные функции, разнеси логически по файлам, все handle функции должны лежать в handl'ах
Все работа с бд должна быть отделена от handle
я вижу тут много парса, вынеси в отдельную дирректорию parse и там напиши необхожимые функкции, пусть даже они будут в две строчки

lilpipidron commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
-		log.Fatal(err)
+		fmt.Println("bcrypt err:", err)
+		return false
 	}
```
думаю, проверку есть ли юзер, можно вынести в отедльную функцию для удобства и упрощения дэбага

lilpipidron commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+		return false
 	}
+	return true
 }
```
чет очень сложно выглядит, мне кажется мой вариант был бы лучше

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
 	}
-	_, err = db.Exec(`INSERT INTO logindata(hashlogin, hashpassword) VALUES ($1, $2)`, hex.EncodeToString(loginHash[:]), hex.EncodeToString(passwordHash[:]))
+
+	// провреяем есть ли пользователь
```
Перевести тут и в остальных местах на английский коммент

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+	for _, char := range password {
+		switch {
+
+		// есть прописные буквы
```
Тут и без комментариев очевидно что делается, убери их

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+		}
+	}
+
+	// проверка длины имени пользлвателя
```
Вынеси магические числа в константы

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+
+	// возвращаем на регистрацию, если не так
+	if !nameLength || !nameAlphaNumeric {
+		tpl.ExecuteTemplate(w, "register.html", "Пожалуйства проверьте критерии для фамилии пользователя")
```
"пожалуйста,"

zzucch replied:
обработай ошибку

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+	// проверка на то, что в имени только буквы
+	var nameAlphaNumeric = true
+	for _, char := range username {
+		if unicode.IsLetter(char) == false {
```
Заменить '== false' на '!' здесь и в остальных местах

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+	if !result {
+		tpl.ExecuteTemplate(w, "register.html", "Такое имя пользователя уже есть")
+	}
+	fmt.Fprint(w, "congrats, your account has been successfully created")
```
'congrats' -> congratulations?

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+
+
+func registerHandler(w http.ResponseWriter, r *http.Request) {
+	fmt.Println("*****registerHandler running*****")
```
может быть тут и в остальных местах будет лучше использовать log.Print а не fmt.Print?

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+
+func registerHandler(w http.ResponseWriter, r *http.Request) {
+	fmt.Println("*****registerHandler running*****")
+	tpl.ExecuteTemplate(w, "main.html", nil)
```
обработай ошибку

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+}
+
+func registerAuthHandler(w http.ResponseWriter, r *http.Request) {
+
```
убери пустую строку

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+
+func registerAuthHandler(w http.ResponseWriter, r *http.Request) {
+
+	r.ParseForm()
```
обработай ошибку

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+
+	// возвращаем на регистрацию, если не так
+	if !nameLength || !nameAlphaNumeric {
+		tpl.ExecuteTemplate(w, "register.html", "Пожалуйства проверьте критерии для имени пользователя")
```
обработай ошибку

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+		return
+	}
+
+	usersurname := r.FormValue("usersurname")
```
usersurname -> surname

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+
+	// возвращаем на регистрацию, если не так
+	if !nameLength || !nameAlphaNumeric {
+		tpl.ExecuteTemplate(w, "register.html", "Пожалуйства проверьте критерии для логина")
```
обработай ошибку

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+	password := r.FormValue("password")
+	fmt.Println("password:", password, "\npswdLength:", len(password))
+
+	var pswdLowercase, pswdUppercase, pswdNumber, pswdSpecial, pswdLength, pswdNoSpaces bool
```
может быть будет лучше сделать pswd -> pwd?

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+
+	// если не выполнены условия возвращаемся на регистрацию
+	if !pswdLowercase || !pswdUppercase || !pswdNumber || !pswdSpecial || !pswdLength || !pswdNoSpaces {
+		tpl.ExecuteTemplate(w, "register.html", "Пожалуйства проверьте критерии для пароля")
```
обработай ошибку

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+	result := AddUser(username, usersurname, login, password)
+	// если пользователь уже есть, возвращаем на регистрацию
+	if !result {
+		tpl.ExecuteTemplate(w, "register.html", "Такое имя пользователя уже есть")
```
обработай ошибку

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
-	if err != nil {
-		log.Fatal(err)
+var tpl *template.Template
+var db *sql.DB
```
не используешь, а потом shadow-ишь на строке 142

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+	row := db.QueryRow(stmt, login)
+	var uID string
+	err = row.Scan(&uID)
+	if err != sql.ErrNoRows {
```
замени '!=' на 'errors.Is()'

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+	db, err = sql.Open("mysql", "root:FuFa2020@tcp(127.0.0.1:3306)/HistoryProject")
 	if err != nil {
-		log.Fatal(err)
+		fmt.Println("Openning DB Error:")
```
openning -> opening

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,188 @@
+		return false
 	}
-	_, err = db.Exec(`DELETE FROM userdata WHERE hashlogin = $1`, hex.EncodeToString(loginHash[:]))
+	defer insertStmt.Close()
```
обработай ошибку

aleksandra0KR committed:
Add files via upload

fix: registration is divided into files, comments are translated, conclusions are removed to the console, insertion is simplified

aleksandra0KR committed:
Update database.go

fix: database file is divided into several files

aleksandra0KR committed:
Add files via upload

fix: parsing in it's own file

lilpipidron commented on internal/database/registerAuthHandler.go:
```diff
@@ -0,0 +1,121 @@
+
+		case unicode.IsLower(char):
+			pswdLowercase = true
+
+		case unicode.IsUpper(char):
+			pswdUppercase = true
+
+		case unicode.IsNumber(char):
+			pswdNumber = true
+
+		case unicode.IsPunct(char) || unicode.IsSymbol(char):
+			pswdSpecial = true
+
+		case unicode.IsSpace(int32(char)):
+			pswdNoSpaces = false
+		}
+	}
+
+	if 11 < len(password) && len(password) < 60 {
+		pswdLength = true
+	}
+
+	if !pswdLowercase || !pswdUppercase || !pswdNumber || !pswdSpecial || !pswdLength || !pswdNoSpaces {
+		err := tpl.ExecuteTemplate(w, "register.html", "Пожалуйста, проверьте критерии для пароля")
+		if err != nil {
+			log.Fatal(err)
+		}
+		return
+	}
+
+	result := AddUser(username, usersurname, login, password)
+
+	// check if the person is already in the database
+	if !result {
+		err := tpl.ExecuteTemplate(w, "register.html", "Такое имя пользователя уже есть")
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+	fmt.Fprint(w, "congrats, your account has been successfully created")
```
Валидацию в другой файлик
validators/validation.go
DRY мне показался тут имеется в твоих проверках

aleksandra0KR committed:
Add files via upload

fix: validation of user information in separated files

aleksandra0KR committed:
Update registerAuthHandler.go

fix: changed registerAuthHandler.go because of the separate validation

zzucch commented on internal/database/AddUser.go:
```diff
@@ -0,0 +1,31 @@
+
+	db, err := sql.Open("mysql", "root:FuFa2020@tcp(127.0.0.1:3306)/HistoryProject")
+	if err != nil {
+		panic(err.Error())
```
убери панику

zzucch commented on internal/database/database.go:
```diff
@@ -1,41 +1,23 @@
 	if err != nil {
-		log.Fatal(err)
+		fmt.Println("Parsing Templates Error:")
+		panic(err.Error)
```
панику убери

aleksandra0KR committed:
Update AddUser.go

fix: deleted panic from AddUser function

aleksandra0KR committed:
Update database.go

fix: deleted panic from startregistration function
